MyOnboarding/src/MyOnboardingApp.Api/Controllers/TodoListController.cs

```diff
@@ -15,45 +16,206 @@ namespace MyOnboardingApp.Api.Controllers
         {
-            await _repository.ReplaceItemAsync(item);
+            ValidateItemBeforePut(replacingItem, id);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var existingItem = await _retrieveService.GetItemByIdAsync(replacingItem.Id);
+            if (!existingItem.WasOperationSuccessful)
+            {
+                var itemToAdd = PrepareItemForCreation(replacingItem);
+                return await PostAsync(itemToAdd);
+            }
+
+            var completedItem = CompleteItemBeforePut(replacingItem, existingItem.Item);
+            var editedItem = await _editService.EditItemAsync(completedItem);
+
+            // ReSharper disable once InvertIf
+            if (!editedItem.WasOperationSuccessful)
+            {
+                foreach (var error in editedItem.Errors)
+                {
+                    ModelState.AddModelError(error.Location, error.Message);
+                }
+
+                return BadRequest(ModelState);
+            }
+
             return StatusCode(HttpStatusCode.NoContent);
         }
-            
+
 
         [Route("{id}")]
-        public async Task<IHttpActionResult> DeleteAsync(Guid id) 
-            => Ok(await _repository.DeleteItemAsync(id));
+        public async Task<IHttpActionResult> DeleteAsync(Guid id)
+        {
+            ValidateNonEmptyGuid(id);
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
+            }
+
+            var deletedItemWithStatus = await _deleteService.DeleteItemAsync(id);
+            if (!deletedItemWithStatus.WasOperationSuccessful)
+            {
+                return StatusCode(HttpStatusCode.NoContent);
+            }
+
+            return Ok(deletedItemWithStatus.Item);
+        }
+
+
+        private static TodoListItem PrepareItemForCreation(TodoListItem item)
+        {
+            item.Id = Guid.Empty;
+            return item;
+        }
+
+
+        private TodoListItem CompleteItemBeforePut(TodoListItem replacingItem, TodoListItem existingItem)
+        {
+            replacingItem.CreationTime = existingItem.CreationTime;
+            replacingItem.LastUpdateTime = existingItem.LastUpdateTime;
+            return replacingItem;
+        }
+
+
+        private void ValidateNonEmptyGuid(Guid id)
+        {
+            if (id == Guid.Empty)
+            {
+                ModelState.AddModelError(nameof(id), "Id must not be empty.");
+            }
+        }
+
+
+        private void ValidateIdPost(Guid id)
+        {
+            if (id != Guid.Empty)
+            {
+                ModelState.AddModelError(nameof(id), "Id must be empty.");
+            }
+        }
+
+
+        private void ValidateTimePost(DateTime time, string field)
+        {
+            if (time != DateTime.MinValue)
+            {
+                ModelState.AddModelError(field, "Time is supposed to be DateTime.MinValue.");
```

Comment:
```suggestion
                ModelState.AddModelError(field, $"Time is supposed to be {DateTime.MinValue}.");
```
Your API users mustn't be C#-savvy at all and stating some random strongly-typed gibberish might frighten them away 😆